src/plugins/thirdParty/Python/src/pythonplugin.cpp

```diff
@@ -0,0 +1,309 @@
+#endif
+
+//==============================================================================
+
+namespace OpenCOR {
+namespace Python {
+
+//==============================================================================
+
+PLUGININFO_FUNC PythonPluginInfo()
+{
+    Descriptions descriptions;
+
+    descriptions.insert("en", QString::fromUtf8("a plugin to provide the <a href=\"https://www.python.org/\">Python</a> language."));
+    descriptions.insert("fr", QString::fromUtf8("une extension ..."));
+
+    return new PluginInfo(PluginInfo::ThirdParty, false, true,
+                          QStringList() << "zlib" << "OpenSSL",
+                          descriptions);
+}
+
+//==============================================================================
+// CLI interface
+//==============================================================================
+
+int PythonPlugin::executeCommand(const QString &pCommand,
+                                        const QStringList &pArguments)
+{
+    // Pass the given CLI command to the Python interpreter
+
+    if (!pCommand.compare("help")) {
+        // Display the commands that we support
+
+        runHelpCommand();
+
+        return 0;
+    } else if (pCommand.isEmpty() || !pCommand.compare("shell")) {
+        // Run the Python interpreter
+
+        return runPythonShell(pArguments);
+    } else {
+        // Not a CLI command that we support, so show our help and leave
+
+        runHelpCommand();
+
+        return -1;
+    }
+}
+
+//==============================================================================
+// Plugin interface
+//==============================================================================
+
+bool PythonPlugin::definesPluginInterfaces()
+{
+    // We don't handle this interface...
+
+    return false;
+}
+
+//==============================================================================
+
+bool PythonPlugin::pluginInterfacesOk(const QString &pFileName,
+                                      QObject *pInstance)
+{
+    Q_UNUSED(pFileName);
+    Q_UNUSED(pInstance);
+
+    // We don't handle this interface...
+
+    return false;
+}
+
+//==============================================================================
+
+void PythonPlugin::initializePlugin()
+{
+    // We must set PYTHONHOME to where Python is 'installed' **before** calling
+    // any Python library code. Calling `Py_SetPythonHome()` doesn't work as
+    // the `Py_Initialize()` code first checks the environment and, if PYTHONHOME
+    // isn't set, uses the installation path compiled in at Python build time...
+
+    QStringList applicationDirectories = QCoreApplication::applicationDirPath().split("/");
+    applicationDirectories.removeLast();
+
+    QString pythonHome = QString();
+#if __APPLE__
+    pythonHome = (applicationDirectories << "Frameworks" << "Python").join("/");
+#elif _WIN32
+    pythonHome = (applicationDirectories << "Python").join("/");
+#else
+    pythonHome = applicationDirectories.join("/");
```

Comment:
I understand that for macOS, you put Python in a folder of its own under `Frameworks`, but when it comes to Linux, why don't you do the same as on Windows?

Reply:
Mainly because having Python as part of the generic `bin`, 'include`, 'lib` directory structure is natural for Linux but that's not the case for Windows (where each installed application has its own set of directories). I'm happy to have Python in its own directory for Linux if you think that's a better fit.

Reply:
At this stage, I am still trying to understand what gets deployed and this is where I am a bit confused. It looks like the same thing is deployed on Windows and macOS, but on Linux I can only see Python as a library, i.e. no header files (actually why deploy them on Windows and macOS?), no scripts, no nothing in fact, or so it seems...? (FWIW, I have created a Python-based snapshot on all three platforms and then checked what gets actually deployed.)

Reply:
You're right, there's no need to deploy the headers. I'll check things later tonight and update the CMake scripts as needed (and generate a PR against your branch).

Reply:
Given that we currently only want to use Python at build time to generate documentation, why not simply exclude all Python related stuff from the distribution packages? For macOS though this would mean building Python outside of the `OpenCOR.app` bundle as otherwise it would be packaged.

The headers turn up since we generate a development version of Python (so other plugins can compile against its library) and Python very much uses a standard `bin`, `include`, `lib` directory structure.

Ideally the packaging mechanism would support excluding named files and folders, which I believe CPack doesn't do.

Reply:
I've seen how to do it.
1) Don't copy any EXTERNAL_BINARIES/DIRECTORIES into the respective plugins.
2) Define PYTHON_EXECUTABLE etc. to point to executables in the `bin` and `Scripts` directories of the package's `ext` directory.
3) Remove the CLI interface from the Python plugin.

Then the non-debug Python can be used in a Windows debug build since all we are doing is running the executable, not linking any libraries into OpenCOR.

Expect a pull request.

Reply:
Thanks @dbrnz, looking forward to the PR. However, I am not sure I understand what you mean by your third point. Does it mean that the Python plugin will still get deployed? Anyway, I would probably stick to what you currently have (so that what I have started cleaning up doesn't get lost), but make sure that nothing that is Python-related gets deployed. If *that* could be done, then it would be great.

Reply:
What you've done cleaning up the Python plugin's CLI interface would just stay on its branch (in fact I'll merge things into my main Python branch). The plan is to not deploy any Python-related code but just the dummy Python and PythonPackages plugins (which OpenCOR then won't load since they won't be needed).

All's going well -- just have to resolve how to best set `PYTHONPATH` to pick-up code installed when building PythonPackages (it's no longer copied into a common place). However I'm starting to run into other demands on my time for today so the PR is now not likely to happen until this evening...

Reply:
Hmm, that means that those dummy Python-based plugins would still get deployed then? If so, it's not really clean. Also, I am not a fan of having many branches (I know you have heaps... :)), so I want to delete my [`agarny/issue1607`](https://github.com/agarny/opencor/tree/issue1607) branch at some point (once I will have merged its contents in to the `master` branch).

No worries about the timing of your PR. I can clearly do other things in the meantime.

Reply:
An update -- the dummy plugins won't get deployed if the `add_plugin` macro is not invoked (and there is also no need for a `src` directory). Now I'm playing with CMake targets and dependencies so that the Python code is actually built/downloaded...